### PR TITLE
Add biometry type detection to local_auth plugin

### PR DESCRIPTION
**Pull request by YAPPDEF**
Should resolve Issue https://github.com/flutter/flutter/issues/22727 

Two new functions were added to the **local_auth** plugin
* `bool canCheckBiometrics` > if device is able to do local auth (with biometrics)
* `List<BiometricType> getAvailableBiometrics` > list of **enrolled** biometrics

Enum BiometricType:
* face
* fingerprint
* _iris (not yet implemented)_

@mehmetf here's the fix according to your mail. Thanks!

### CODE REVIEW

File: packages/local_auth/ios/Classes/LocalAuthPlugin.m [hunk 1]
```diff
@@ -62,6 +64,25 @@ - (void)alertMessage:(NSString *)message
...
+- (void)getBiometryType:(FlutterResult)result {
+    LAContext *context = [[LAContext alloc] init];
+    NSError *authError = nil;
+    if ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&authError]) {
+        if (authError == nil) {
+            if (@available(iOS 11.0.1, *)) {
+                if (context.biometryType == LABiometryTypeFaceID) {
+                    result(@"faceID");
+                } else if (context.biometryType == LABiometryTypeTouchID) {
+                    result(@"touchID");
```
Comment (mehmetf) [inline packages/local_auth/ios/Classes/LocalAuthPlugin.m]: Let's not use platform specific constants.

If you want to show platform specific messaging you can do the following in your app:

```dart
if (Platform.isIOS && getAvailableBiometrics.contains(FINGERPRINT)) {
   // this is Touch ID.
}
```

File: packages/local_auth/ios/Classes/LocalAuthPlugin.m [hunk 2]
```diff
@@ -62,6 +64,25 @@ - (void)alertMessage:(NSString *)message
...
 }
 
+- (void)getBiometryType:(FlutterResult)result {
+    LAContext *context = [[LAContext alloc] init];
+    NSError *authError = nil;
+    if ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&authError]) {
+        if (authError == nil) {
+            if (@available(iOS 11.0.1, *)) {
+                if (context.biometryType == LABiometryTypeFaceID) {
+                    result(@"faceID");
```
Comment (mehmetf) [inline packages/local_auth/ios/Classes/LocalAuthPlugin.m]: Let's not use platform specific constants. Use "face".

File: packages/local_auth/lib/local_auth.dart [hunk 3]
```diff
@@ -72,4 +72,13 @@ class LocalAuthentication {
     }
     return await _channel.invokeMethod('authenticateWithBiometrics', args);
   }
+
+  /// Returns biometry type if enrolled on the device
+  ///
+  /// Returns a [Future] String with the following possibilities:
+  /// - iOS: 'faceID', 'touchID', 'notEnrolled' or 'notAvailable'
+  /// - Android: 'fingerprint', 'notEnrolled' or 'notAvailable'
+  Future<String> getBiometryType() async {
```
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: Map the returned string to an enum instead so that it is easier for applications to use your API (also see the suggested API in my first comment).

File: packages/local_auth/ios/Classes/LocalAuthPlugin.m (hunk 1 above)
Comment (YAPPDEF) [reply packages/local_auth/ios/Classes/LocalAuthPlugin.m]: If I get error.code = LAErrorTouchIDNotEnrolled, the differentiation between touchID and faceID is not yet available (not yet enrolled). 

In this case, should I return ["undefined"] to detect that in dart that it would be possible (canCheckBiometrics=true) but not yet which type (getAvailableBiometrics=['undefined'])?

**Comment by YAPPDEF** [comment]
I signed it!

File: packages/local_auth/example/lib/main.dart [hunk 4]
```diff
@@ -17,8 +17,40 @@ class MyApp extends StatefulWidget {
 }
 
 class _MyAppState extends State<MyApp> {
+  bool _canCheckBiometrics;
+  List<BiometricType> _availableBiometrics;
   String _authorized = 'Not Authorized';
 
+  Future<Null> _checkBiometrics() async {
+    final LocalAuthentication auth = LocalAuthentication();
```
Comment (mehmetf) [inline packages/local_auth/example/lib/main.dart]: You could just create a single _auth object for the entire class and reuse it.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java [hunk 5]
```diff
@@ -72,6 +74,16 @@ public void onError(String code, String error) {
                 }
               });
       authenticationHelper.authenticate();
+    } else if (call.method.equals("getAvailableBiometrics")) {
+      FingerprintManager fingerprintMgr =
+          registrar.activity().getSystemService(FingerprintManager.class);
+      ArrayList<String> biometrics = new ArrayList<String>();
+      if (!fingerprintMgr.hasEnrolledFingerprints()) {
+        biometrics.add("undefined");
+      } else if (fingerprintMgr.isHardwareDetected() && fingerprintMgr.hasEnrolledFingerprints()) {
```
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: `isHardwareDetected` is a device capability call. You should probably check that first:

```java
if (isHardwareDetected) {
  if (hasEnrolledFingerprint) {
    biometrics.add("fingerprint");
  } else {
    biometrics.add("undefined");
  }
}
result.success(biometrics);
```

File: packages/local_auth/ios/Classes/LocalAuthPlugin.m [hunk 6]
```diff
@@ -62,6 +64,26 @@ - (void)alertMessage:(NSString *)message
                                                                                    completion:nil];
 }
 
+- (void)getAvailableBiometrics:(FlutterResult)result {
+  LAContext *context = [[LAContext alloc] init];
+  NSError *authError = nil;
```
Comment (mehmetf) [inline packages/local_auth/ios/Classes/LocalAuthPlugin.m]: Collect the results in a list and return them like you do in java code rather than returning directly. See https://developer.apple.com/documentation/foundation/nsmutablearray .

File: packages/local_auth/lib/local_auth.dart [hunk 7]
```diff
@@ -72,4 +74,40 @@ class LocalAuthentication {
     }
     return await _channel.invokeMethod('authenticateWithBiometrics', args);
   }
+
+  /// Returns true or false if biometrics are available
```
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: Fix comment. This checks if device is capable of checking biometrics not whether biometrics is available. You can have a device with biometric capability but not enrolled in any. In the latter case, this returns true (but no biometrics is available for the app to check).

File: packages/local_auth/lib/local_auth.dart [hunk 8]
```diff
@@ -72,4 +74,40 @@ class LocalAuthentication {
...
+  }
+
+  /// Returns a list of enrolled biometrics
+  ///
+  /// Returns a [Future] List<BiometricType> with the following possibilities:
+  /// - BiometricType.face
+  /// - BiometricType.fingerprint
+  /// - BiometricType.iris (not yet implemented)
+  Future<List<BiometricType>> getAvailableBiometrics() async {
+    final List<dynamic> result =
```
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: List\<string\> instead of dynamic.

File: packages/local_auth/lib/local_auth.dart [hunk 9]
```diff
@@ -72,4 +74,40 @@ class LocalAuthentication {
...
+        await _channel.invokeMethod('getAvailableBiometrics');
+
+    final List<BiometricType> biometrics = <BiometricType>[];
+    result.forEach((dynamic value) {
```
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: string value

File: packages/local_auth/lib/local_auth.dart [hunk 10]
```diff
@@ -72,4 +74,40 @@ class LocalAuthentication {
...
+        case 'face':
+          biometrics.add(BiometricType.face);
+          break;
+        case 'fingerprint':
+          biometrics.add(BiometricType.fingerprint);
+          break;
+        case 'iris':
+          biometrics.add(BiometricType.iris);
+          break;
+      }
```
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: handle 'undefined' as well (in which case you won't do anything).

File: packages/local_auth/lib/local_auth.dart (hunk 8 above)
Comment (YAPPDEF) [reply packages/local_auth/lib/local_auth.dart]: Throws an exception, that's why I use dynamic
```
[VERBOSE-2:shell.cc(181)] Dart Error: Unhandled exception:
type 'List<dynamic>' is not a subtype of type 'List<String>'
#0      LocalAuthentication.getAvailableBiometrics (package:local_auth/local_auth.dart:94:24)
```

File: packages/local_auth/lib/local_auth.dart (hunk 9 above)
Comment (YAPPDEF) [reply packages/local_auth/lib/local_auth.dart]: Same as before

File: packages/local_auth/lib/local_auth.dart (hunk 8 above)
Comment (mehmetf) [reply packages/local_auth/lib/local_auth.dart]: That's an unfortunate artifact of the plugin system as it does not allow you to generify the methods easily. You can explicitly cast it:

```dart
  (await _channel.invokeMethod(...) as List).cast<String>();
```

Comment (YAPPDEF) [reply packages/local_auth/lib/local_auth.dart]: I added the cast (+added type NSString for NSMutableArray)

I omitted `as List` for the cast, as it caused this warning: 
`[dart] Avoid using `as`.`

**Comment by mehmetf** [comment]
Thank you! This looks good now. One last thing:

Please update changelog.md and pubspec.yaml with a new version: `0.3.0` since there's a new API.

File: packages/local_auth/lib/local_auth.dart [hunk 11]
```diff
@@ -72,4 +74,41 @@ class LocalAuthentication {
...
     return await _channel.invokeMethod('authenticateWithBiometrics', args);
   }
+
+  /// Returns true if device is capable of checking biometrics
+  ///
+  /// Returns a [Future] bool true or false:
+  Future<bool> canCheckBiometrics() async {
+    final List<dynamic> biometrics =
+        await _channel.invokeMethod('getAvailableBiometrics');
+    return biometrics.isNotEmpty;
```
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: Can you shorten this to:

```dart
Future<bool> get canCheckBiometrics async => (await _channel.invo.....).isNotEmpty;
```

I would like to avoid putting `dynamic` in this code.

**Comment by mehmetf** [comment]
Done!
